bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieImpl.java

```diff
@@ -112,8 +113,9 @@ public class BookieImpl extends BookieCriticalThread implements Bookie {
 
     private int exitCode = ExitCode.OK;
 
-    private final ConcurrentLongHashMap<byte[]> masterKeyCache =
-            ConcurrentLongHashMap.<byte[]>newBuilder().build();
+    // key: masterKey, values: set of ledgerId
+    private final static ConcurrentOpenHashMap<byte[], ConcurrentLongHashSet> masterKeyCache
```

Comment:
this flips cache from ledgerId -> key to key -> set of ledgerids.
While it reduces memory usage in cases of the same key, it does not help if ledgers are stored (not gc'd) for too long, there are too many ledgers etc. It will slow down rate at which cache eats up the memory. 
It also makes get/remove operations for the cache slow.

What's worse is that outside of Pulsar we don't know how people use the keys, what if some user sets a unique key for each ledger? the problem just got worse for them.

Reply:
@dlg99 
use this structure ConcurrentOpenHashMap, the key is masterKey, the value is the set of ledgerIds,
the masterKey settings are different, 
1. which increases the time-consuming of remove, but the remove operation is timed, not so frequent
2. when we add the new data ,we can use `containsMasterKeyCache` to spead up weather to `addMasterKeyCache`